src/botPage/bot/TradeEngine/index.js

```diff
@@ -20,15 +20,26 @@ import { start } from './state/actions';
 const watchBefore = store =>
     watchScope({
         store,
-        stopScope: constants.DURING_PURCHASE,
-        passScope: constants.BEFORE_PURCHASE,
-        passFlag : 'proposalsReady',
+        stopScope    : constants.DURING_PURCHASE,
+        passScope    : constants.BEFORE_PURCHASE,
+        passCondition: state => {
+            if (state.proposalsReady) {
+                return true;
+            }
+            return false;
+        },
     });
 
 const watchDuring = store =>
-    watchScope({ store, stopScope: constants.STOP, passScope: constants.DURING_PURCHASE, passFlag: 'openContract' });
+    watchScope({
+        store,
+        stopScope    : constants.STOP,
+        passScope    : constants.DURING_PURCHASE,
+        passCondition: state => state.openContract,
+    });
 
-const watchScope = ({ store, stopScope, passScope, passFlag }) => {
+let prevTick;
+const watchScope = ({ store, stopScope, passScope, passCondition }) => {
```

Comment:
It's better to move `prevTick` to inside `watchScope`, otherwise, it's polluting the module scope and the `prevTick` from one scope might interfere with other non-related scope (user can miss one tick for `before` scope)